EIPS/eip-5164.md

```diff
@@ -26,203 +26,286 @@ The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL
 
 This specification allows contracts on one chain to send messages to contracts on another chain. There are two key interfaces that needs to be implemented:
 
-- `CrossChainRelayer`
-- `CrossChainExecutor`
+- `MessageDispatcher`
+- `MessageExecutor`
 
-The `CrossChainRelayer` lives on the origin chain. Users can share a single `CrossChainRelayer` or deploy their own.
+The `MessageDispatcher` lives on the origin chain and dispatches messages to the `MessageExecutor` for execution. The `MessageExecutor` lives on the destination chain and executes dispatched messages.
 
-The `CrossChainExecutor` lives on the destination chain and executes relayed calls. Users can share a single `CrossChainExecutor` or deploy their own.
+There are also extensions of `MessageExecutor`, each defining a method for initiating a message or message batch:
 
-### CrossChainRelayer
+- `SingleChainMessageDispatcher`
+- `MultiChainMessageDispatcher`
+- `BatchedMessageDispatcher`
 
-The `CrossChainRelayer` lives on the chain from which messages are sent. The Relayer's job is to broadcast calls through a transport layer to one or more `CrossChainExecutor` contracts.
+Alternatively, `MessageDispatcher`s may implement a custom interface for initiating messages.
 
-#### Methods
+### MessageDispatcher
+
+The `MessageDispatcher` lives on the chain from which messages are sent. The Dispatcher's job is to broadcast messages through a transport layer to one or more `MessageExecutor` contracts.
+
+#### MessageDispatcher Events
 
-**relayCalls**
+A unique `messageId` MUST be generated for each message or message batch.
```

Comment:
I think we should lift this up into the MessageDispatcher description, and give some more detail.  It's going to be important for the messageId to be unique across contracts and chains- perhaps we can suggest that it's a monotonically increasing nonce